### PR TITLE
refactor(challenger): support three dispute paths with game classification

### PR DESCRIPTION
**Pull request by jackchuma**
## Summary

Refactors the challenger to classify dispute games into three distinct paths and dispatch to the appropriate handler, replacing the previous single-path nullification-only flow.

### Three dispute paths

1. **Wrong TEE proof** (`InvalidTeeProposal`) — nullify with a TEE proof or challenge with a ZK proof via `challenge()`.
2. **Fraudulent ZK challenge** (`FraudulentZkChallenge`) — a correct TEE proposal was challenged with a wrong ZK proof; nullify the fraudulent challenge with a ZK proof via `nullify()`.
3. **Wrong ZK proposal** (`InvalidZkProposal`) — nullify the incorrect ZK proposal with a ZK proof via `nullify()`.

### Key changes

- **`GameCategory` enum + `classify()` in scanner** — games are now classified based on `teeProver`, `zkProver`, and `counteredByIntermediateRootIndexPlusOne` state. Scanner fetches `countered_index` and `l1_head` alongside existing fields.
- **`DisputeIntent` enum** (`Challenge` | `Nullify`) — attached to each `PendingProof` to drive which on-chain function is called at submission time.
- **Driver dispatch** — `process_candidate` routes to `process_invalid_proposal` (Paths 1 & 3) or `process_fraudulent_zk_challenge` (Path 2). Validation logic extracted into `validate_game`.
- **`poll_or_submit` improvements** — intent-aware liveness checks: Challenge intents verify `teeProver` is still set; Nullify intents check whether the targeted prover has already been zeroed, preventing infinite retry loops.
- **Submitter** — `submit_nullification` → `submit_dispute`, which encodes either `nullify()` or `challenge()` calldata based on `DisputeIntent`.
- **TEE proof sourcing** — uses the game's stored L1 head (from CWIA) instead of fetching the current finalized head; `L1HeadProvider` trait changed from `finalized_head()` to `block_number_by_hash()`. Proof encoding switched to compact format (`encode_dispute_proof_bytes` — type + signature only, no L1 origin data).
- **Contracts** — added `encode_challenge_calldata`, `encode_dispute_proof_bytes`, `countered_index`, and `l1_head` to `AggregateVerifierClient`.
- **Metrics** — new counters for challenge tx submitted/outcome/latency, fraudulent ZK challenge detected, and invalid ZK proposal detected.
- **Tests** — significantly expanded driver and submitter tests covering all three paths, including fraudulent challenge validation edge cases.

### CODE REVIEW

**Comment by cb-heimdall** [comment]
<h1>✅ Heimdall Review Status</h1>
<table>
<tr>
<th>Requirement</th>
<th>Status</th>
<th>More Info</th>
</tr>
<tr>
<td>Reviews</td>
<td>
✅
<code>1/1</code>
</td>
<td>
<details>
<summary>Denominator calculation</summary>
<table>
<tr>
<td>
<details>
<summary>Show calculation</summary>
<table>
<tr>
<td><code>1</code> if user is bot</td>
<td>0</td>
</tr>
<tr>
<td><code>1</code> if user is external</td>
<td>0</td>
</tr>
<tr>
<td><code>2</code> if repo is sensitive</td>
<td>0</td>
</tr>
<tr>
<td>
From <code>.codeflow.yml</code>
</td>
<td>1</td>
</tr>
<tr>
<td>
Additional review requirements
<details>
<summary>Show calculation</summary>
<table>
<tr>
<td><strong>Max</strong></td>
<td><strong>0</strong></td>
</tr>
</table>
</details>
</td>
<td>0</td>
</tr>
<tr>
<td>
From <code>CODEOWNERS</code>
</td>
<td>0</td>
</tr>
<tr>
<td>Global minimum</td>
<td>0</td>
</tr>
<tr>
<td><strong>Max</strong></td>
<td>
<strong>
1
</strong>
</td>
</tr>
</table>
</details>
</td>
<td>1</td>
</tr>
<tr>
<td><code>1</code> if commit is unverified</td>
<td>0</td>
</tr>
<tr>
<td><strong>Sum</strong></td>
<td><strong>1</strong></td>
</tr>
</table>
</details>
</td>
</tr>
</table>

File: crates/proof/challenge/src/driver.rs [hunk 1]
```diff
@@ -439,32 +574,84 @@ impl<L2: L2Provider, P: ZkProofProvider, T: TxManager> Driver<L2, P, T> {
...
-            self.pending_proofs.remove(&game_address);
-            return Ok(());
-        }
-        if zk_prover != Address::ZERO {
-            debug!(game = %game_address, zk_prover = %zk_prover, "game already challenged, dropping pending proof");
-            self.pending_proofs.remove(&game_address);
-            return Ok(());
+        let (invalid_index, expected_root, intent, targets_tee) =
+            match self.pending_proofs.get(&game_address) {
+                Some(p) => (p.invalid_index, p.expected_root, p.intent, p.prove_request.is_none()),
```
Comment (leopoldjoy) [inline crates/proof/challenge/src/driver.rs:589]: `targets_tee = prove_request.is_none()` is a fragile proxy for "targets TEE prover slot" — would it be viable to store an explicit `ProofKind` field on `PendingProof` instead of inferring it from the absence of a ZK request? (Particularly when considering how this codebase may evolve in the future as proof types may change, etc.)

File: crates/proof/challenge/src/scanner.rs [hunk 2]
```diff
@@ -193,22 +241,22 @@ impl GameScanner {
...
-            debug!(
-                index = index,
-                zk_prover = %zk_prover,
-                "skipping already-challenged game"
-            );
-            return Ok(None);
-        }
-
-        let (info, starting_block_number, tee_prover) = tokio::try_join!(
+        let (zk_prover, tee_prover, countered_index, info, starting_block_number, l1_head) = tokio::try_join!(
```
Comment (leopoldjoy) [inline crates/proof/challenge/src/scanner.rs:244]: Nit: all 6 RPC calls are issued before classification, wasting calls on games that classify will skip; consider fetching only (`zk_prover`, `tee_prover`, `countered_index`) first, classifying, then fetching the rest.

File: crates/proof/challenge/src/driver.rs [hunk 3]
```diff
@@ -256,34 +307,125 @@ impl<L2: L2Provider, P: ZkProofProvider, T: TxManager> Driver<L2, P, T> {
...
+        // challenge was legitimate — skip.
+        //
+        // The validator scans intermediate roots sequentially and reports the
+        // first invalid index. If `first_invalid <= challenged_index`, the
+        // root at the challenged index (or an earlier one) was wrong, so the
+        // ZK challenge was legitimate.
+        if !result.is_valid {
+            match result.invalid_intermediate_index {
+                Some(first_invalid)
+                    if u64::try_from(first_invalid).unwrap_or(u64::MAX) <= challenged_index =>
```
Comment (leopoldjoy) [inline crates/proof/challenge/src/driver.rs:351]: Nit: `unwrap_or(u64::MAX)` silently swallows a `usize` -> `u64` conversion error; propagate the error or at minimum log it, since this comparison determines whether a fraudulent challenge is skipped.

**Review by leopoldjoy — APPROVED** [review]
None of the above is blocking, can be addressed later
